keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -0,0 +1,248 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Manage the vendor folder in `kubernetes/kubernetes` using go modules,
+and define `go.mod` module files for published components like `k8s.io/client-go` and `k8s.io/api`.
+
+## Motivation
+
+Since its inception, Kubernetes has used Godep to manage vendored 
+dependencies, to ensure reproducible builds and auditable source.
+
+As the go ecosystem matured, vendoring became a first-class concept,
+Godep became unmaintained, Kubernetes started using a custom version of Godep,
+other vendoring tools (like `glide` and `dep`) became available,
+and dependency management was ultimately added directly to go in the form of go modules.
+
+The [plan of record](https://blog.golang.org/modules2019) is for go1.13 to enable
+go modules by default and deprecate GOPATH mode. To be ready for that, Kubernetes
+components must make any required adjustments in the Kubernetes 1.15 release.
+
+In addition to simply keeping up with the go ecosystem, go modules provide many benefits:
+* rebuilding `vendor` with go modules provided a 10x speed increase over Godep in preliminary tests
+* go modules can reproduce a consistent `vendor` directory on any OS
+* if semantic import versioning is adopted, consumers of Kubernetes modules can use two distinct versions simultaneously (if required by diamond dependencies)
+
+### Goals
+
+* Remove the kubernetes/kubernetes dependency on godep (including the need for a customized version of godep)
+* Increase speed and reliability of building and verifying the `vendor` directory in kubernetes/kubernetes
+* Provide accurate go.mod module descriptors for kubernetes/kubernetes and published staging components
+* Enable use of published kubernetes components by `go module` aware consumers
+* Allow continued use of published kubernetes components by `go module` *unaware* consumers until go modules are enabled by default
+
+## Proposal
+
+### Manage vendor folders using go modules
+1. Make `k8s.io/kubernetes` and each staging component (like `k8s.io/client-go` and `k8s.io/api`) a distinct go module
+  * Pin dependencies to the same versions that are currently listed in Godeps.json
+2. Change vendor management and verification scripts in `kubernetes/kubernetes` to use go module commands
+3. Change the staging component publishing bot to modify the `go.mod` pinned versions of peer components, rather than rewriting Godeps.json files
+
+### Select a versioning strategy for published modules
```

Comment:
The working area for this could be the code organization subproject of SIG Arch. That is, if there is need to work through the strategy.

Reply:
After quite a bit of investigation, I think the best approach would be to start small and keep the current tagging strategy for the first iteration and consider changes to our component versioning separately. Moved the considered strategies to the alternatives section to preserve the documented impact.